psydac/linalg/tests/test_linalg.py

```diff
@@ -0,0 +1,959 @@
+    U = BlockVectorSpace(V, W)
+    B = BlockLinearOperator(U, U, ((S, None), (None, S1)))
+    u = BlockVector(U, (v,w))
+
+    ###
+    ### BlockLO Transpose - Inverse Tests
+    ### -1,T & T,-1 --- -1,T,T --- -1,T,-1 --- T,-1,-1 --- T,-1,T (the combinations I test)
+    ###
+
+    tol = 1e-5
+    C = inverse(B, 'cg', tol=tol)
+
+    # -1,T & T,-1 -> equal
+    assert isinstance(C.T, ConjugateGradient)
+    assert isinstance(inverse(B.T, 'cg', tol=tol), ConjugateGradient)
+    assert np.sqrt(sum(((C.T.dot(u) - inverse(B.T, 'cg', tol=tol).dot(u)).toarray())**2)) < 2*tol
+    # -1,T,T -> equal -1
+    assert np.sqrt(sum(((C.T.T.dot(u) - C.dot(u)).toarray())**2)) < 2*tol
+    # -1,T,-1 -> equal T
+    assert isinstance(inverse(C.T, 'bicg'), BlockLinearOperator)
+    assert np.all(inverse(C.T, 'bicg').dot(u).toarray() == B.T.dot(u).toarray())
+    # T,-1,-1 -> equal T
+    assert isinstance(inverse(inverse(B.T, 'cg', tol=tol), 'pcg', pc='jacobi'), BlockLinearOperator)
+    assert np.all(inverse(inverse(B.T, 'cg', tol=tol), 'pcg', pc='jacobi').dot(u).toarray() == B.T.dot(u).toarray())
+    # T,-1,T -> equal -1
+    assert isinstance(inverse(B.T, 'cg', tol=tol).T, ConjugateGradient)
+    assert np.sqrt(sum(((inverse(B.T, 'cg', tol=tol).dot(u) - C.dot(u)).toarray())**2)) < tol
+
+    ###
+    ### StencilMatrix Transpose - Inverse Tests
+    ### -1,T & T,-1 --- -1,T,T --- -1,T,-1 --- T,-1,-1 --- T,-1,T (the combinations I test)
+    ###
+
+    tol = 1e-5
+    C = inverse(S, 'cg', tol=tol)
+
+    # -1,T & T,-1 -> equal
+    assert isinstance(C.T, ConjugateGradient)
+    assert isinstance(inverse(S.T, 'cg', tol=tol), ConjugateGradient)
+    assert np.sqrt(sum(((C.T.dot(v) - inverse(S.T, 'cg', tol=tol).dot(v)).toarray())**2)) < 2*tol
+    # -1,T,T -> equal -1
+    assert np.sqrt(sum(((C.T.T.dot(v) - C.dot(v)).toarray())**2)) < 2*tol
+    # -1,T,-1 -> equal T
+    assert isinstance(inverse(C.T, 'bicg'), StencilMatrix)
+    assert np.all(inverse(C.T, 'bicg').dot(v).toarray() == S.T.dot(v).toarray())
+    # T,-1,-1 -> equal T
+    assert isinstance(inverse(inverse(S.T, 'cg', tol=tol), 'pcg', pc='jacobi'), StencilMatrix)
+    assert np.all(inverse(inverse(S.T, 'cg', tol=tol), 'pcg', pc='jacobi').dot(v).toarray() == S.T.dot(v).toarray())
+    # T,-1,T -> equal -1
+    assert isinstance(inverse(S.T, 'cg', tol=tol).T, ConjugateGradient)
+    assert np.sqrt(sum(((inverse(S.T, 'cg', tol=tol).dot(v) - C.dot(v)).toarray())**2)) < tol
+
+#===============================================================================
+#@pytest.mark.parametrize( 'n1', n1array)
+#@pytest.mark.parametrize( 'n2', n2array)
+#@pytest.mark.parametrize( 'p1', p1array)
+#@pytest.mark.parametrize( 'p2', p2array)
+
+# 'cg' inverse requires a positive definite matrix.
+# I did not yet come up with a way to create positive definite matrices for arbitrary n1, n2, p1, p2, P1, P2
+# Thus until changed: Only one test with a simple 4x4 positive definite matrix
```

Comment:
It is actually quite easy: a symmetric strictly diagonally dominant real matrix with positive diagonal entries is positive definite! So it has to be

- real-valued
- symmetric
- with positive entries in the (main) diagonal
- strictly diagonally dominant

Reply:
That's neat, thank you! Will update this test in the near future.